### PR TITLE
Aerogear 1921 validate passing bundle id when creating client

### PR DESCRIPTION
**Pull request by maleck13**
This pr adds a few new things
1) create client API that uses a bundle id
2) set of simple unit tests
3) a interface for running the CLI commands

### CODE REVIEW

File: src/main/java/org/aerogear/plugin/intellij/mobile/api/CliRunnerImpl.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+public class CliRunnerImpl implements CLIRunner {
+  
+  @Override public String runCmd(List<String> cmd, int timeoutInSeconds) throws CLIException {
```
Comment (sedroche) [inline src/main/java/org/aerogear/plugin/intellij/mobile/api/CliRunnerImpl.java]: This method doesn't run within a new thread e.g. `Runnable` which is what is indicated by the class interface naming `CLIRunner`.
Should we keep this class for Runnable API interactions?

File: src/main/java/org/aerogear/plugin/intellij/mobile/api/CliRunnerImpl.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+      outPut = readOutput(p.getInputStream());
+    }catch (IOException e){
+      throw new CLIException("unexpected io error executing cli command : "+e.getMessage(),e.getCause());
+    }catch (InterruptedException e ){
+      throw new CLIException("cli command was unexpectedly interrupted : " + e.getMessage(),e.getCause());
+    }
+    return outPut;
+  }
+  
+  public List<String>prepareCmd(String... args){
```
Comment (sedroche) [inline src/main/java/org/aerogear/plugin/intellij/mobile/api/CliRunnerImpl.java]: This seems more like a private `MobileAPI` method to me. We get a `command` returned and pass it again to `CliRunner`. wdyt?

Comment (sedroche) [inline src/main/java/org/aerogear/plugin/intellij/mobile/api/CliRunnerImpl.java]: More a question on why you used `String... args` here and not `List<String> cmd` as elsewhere?

File: src/main/java/org/aerogear/plugin/intellij/mobile/api/CliRunnerImpl.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    ExecutorService executor = Executors.newFixedThreadPool(1);
+    executor.execute(() -> {
+      try {
+        Process p = pb.start();
+        String input = readOutput(p.getInputStream());
+        String error = readOutput(p.getErrorStream());
+        if (input.length() != 0) {
+          w.onSuccess(input);
+        } else if (error.length() != 0) {
+          w.onError(new Exception(error.toString()));
```
Comment (sedroche) [inline src/main/java/org/aerogear/plugin/intellij/mobile/api/CliRunnerImpl.java]: Should we use `CLIException` everywhere for consistency?

File: src/main/java/org/aerogear/plugin/intellij/mobile/api/MobileAPI.java [hunk 4]
```diff
@@ -1,98 +1,67 @@
...
+        cliRunner.runAndWatch(cmd, w);
+    }
+    
+    
+    public MobileClient createClient(String name, String clientType, String bundleID)throws CLIException{
+        if("".equals(name) || "".equals(clientType) || "".equals(bundleID)){
+            throw new CLIException("expected a client name, a client type and a bundle id");
+        }
+        List<String> cmd = cliRunner.prepareCmd("create","client",name,clientType,bundleID);
+        String res = cliRunner.runCmd(cmd,5);
```
Comment (sedroche) [inline src/main/java/org/aerogear/plugin/intellij/mobile/api/MobileAPI.java]: Can we have a `runCmd(cmd)` that defaults to 5 if not declared?

File: src/main/java/org/aerogear/plugin/intellij/mobile/ui/MobileToolWindowFactory.java [hunk 5]
```diff
@@ -29,7 +33,7 @@ public void createToolWindowContent(Project project, ToolWindow toolWindow){
 
         MobileServices serviceList;
         try {
-            serviceList = new MobileAPI().getServices();
+            serviceList = new MobileAPI(cliRunner).getServices();
```
Comment (sedroche) [inline src/main/java/org/aerogear/plugin/intellij/mobile/ui/MobileToolWindowFactory.java]: What advantages do we get from passing in the cliRunner at creation?

File: src/test/java/MobileAPITest.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+import org.junit.Assert;
+import org.testng.annotations.Test;
+
+import static org.mockito.Mockito.*;
+
+public class MobileAPITest {
+  @Test
+  public void testGetServices_ok() throws Exception {
+    CLIRunner runner = mock(CLIRunner.class);
+    when(runner.runCmd(anyList(),anyInt())).thenReturn("{\n"
```
Comment (sedroche) [inline src/test/java/MobileAPITest.java]: Is it possible to have this as a JSON file we can load or create the JSON using gson with a matching entity?
Just to keep the test cleaner.

File: src/main/java/org/aerogear/plugin/intellij/mobile/api/CliRunnerImpl.java (hunk 1 above)
Comment (maleck13) [reply src/main/java/org/aerogear/plugin/intellij/mobile/api/CliRunnerImpl.java]: we could change it to executeCmd? and in the future add runCmd for background stuff?

File: src/main/java/org/aerogear/plugin/intellij/mobile/api/CliRunnerImpl.java (hunk 2 above)
Comment (maleck13) [reply src/main/java/org/aerogear/plugin/intellij/mobile/api/CliRunnerImpl.java]: yeah I think we can clean this up

File: src/main/java/org/aerogear/plugin/intellij/mobile/ui/MobileToolWindowFactory.java (hunk 5 above)
Comment (maleck13) [reply src/main/java/org/aerogear/plugin/intellij/mobile/ui/MobileToolWindowFactory.java]: allows for passing in the mock during testing also IMHO dependencies should be clear and the constructor is where they should be passed in. I know this isn't always the case with Java as there are quite a few libs for doing dependency injection, but in our case I think it is simple enough to keep it this way for now

File: src/test/java/MobileAPITest.java (hunk 6 above)
Comment (maleck13) [reply src/test/java/MobileAPITest.java]: Debated that internally too. I am not sure I prefer it or not. Part of me likes the explictness of it and also not having to load definitions from files.

Comment (maleck13) [reply src/test/java/MobileAPITest.java]: If you're ok with it, I will leave it as is, and if it becomes a problem lets look to move it out then

**Comment by maleck13** [comment]
@sedroche  believe I have addressed your feedback. Could you take another look

File: src/main/java/org/aerogear/plugin/intellij/mobile/api/CliRunnerImpl.java (hunk 1 above)
Comment (sedroche) [reply src/main/java/org/aerogear/plugin/intellij/mobile/api/CliRunnerImpl.java]: Yep that is more clear to me.

File: src/main/java/org/aerogear/plugin/intellij/mobile/api/CliRunnerImpl.java (hunk 2 above)
Comment (sedroche) [reply src/main/java/org/aerogear/plugin/intellij/mobile/api/CliRunnerImpl.java]: Maybe a `CliCommand` class?

File: src/test/java/MobileAPITest.java (hunk 6 above)
Comment (sedroche) [reply src/test/java/MobileAPITest.java]: Yep, fine with that. i just imagined larger JSON data in some tests but fine for now

File: src/main/java/org/aerogear/plugin/intellij/mobile/ui/MobileToolWindowFactory.java (hunk 5 above)
Comment (sedroche) [reply src/main/java/org/aerogear/plugin/intellij/mobile/ui/MobileToolWindowFactory.java]: :+1: on tests. Hadn't realised

**Pull request merged**